File: app/src/main/java/dev/ragnarok/fenrir/settings/ISettings.java
package dev.ragnarok.fenrir.settings;

import android.content.Context;
import android.net.Uri;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.Set;

import dev.ragnarok.fenrir.AccountType;
import dev.ragnarok.fenrir.api.model.LocalServerSettings;
import dev.ragnarok.fenrir.api.model.PlayerCoverBackgroundSettings;
import dev.ragnarok.fenrir.api.model.SlidrSettings;
import dev.ragnarok.fenrir.crypt.KeyLocationPolicy;
import dev.ragnarok.fenrir.model.Lang;
import dev.ragnarok.fenrir.model.PhotoSize;
import dev.ragnarok.fenrir.model.SideSwitchableCategory;
import dev.ragnarok.fenrir.model.SwitchableCategory;
import dev.ragnarok.fenrir.model.drawer.RecentChat;
import dev.ragnarok.fenrir.place.Place;
import dev.ragnarok.fenrir.settings.theme.ThemeOverlay;
import dev.ragnarok.fenrir.view.pager.Transformers_Types;
import io.reactivex.rxjava3.core.Flowable;
import io.reactivex.rxjava3.core.Observable;

public interface ISettings {

    IRecentChats recentChats();

    IDrawerSettings drawerSettings();

    ISideDrawerSettings sideDrawerSettings();

    IPushSettings pushSettings();

    ISecuritySettings security();

    IUISettings ui();

    INotificationSettings notifications();

    IMainSettings main();

    IAccountsSettings accounts();

    IOtherSettings other();

    interface IOtherSettings {
        String getFeedSourceIds(int accountId);

        void setFeedSourceIds(int accountId, String sourceIds);

        void storeFeedScrollState(int accountId, String state);

        String restoreFeedScrollState(int accountId);

        String restoreFeedNextFrom(int accountId);

        void storeFeedNextFrom(int accountId, String nextFrom);

        boolean isAudioBroadcastActive();

        int getMaxBitmapResolution();

        @NonNull
        List<Integer> getServicePlaylist();

        boolean isNative_parcel_photo();

        boolean isNative_parcel_story();

        boolean isDoLogs();

        boolean isExtra_debug();

        boolean isDump_fcm();

        boolean isAutoplay_gif();

        boolean isStrip_news_repost();

        boolean isAd_block_story_news();

        Set<String> isBlock_news_by_words();

        boolean isNew_loading_dialog();

        String get_Api_Domain();

        String get_Auth_Domain();

        boolean isUse_api_5_90_for_audio();

        boolean isDisable_history();

        boolean isShow_wall_cover();

        boolean isDeveloper_mode();

        boolean isForce_cache();

        boolean isKeepLongpoll();

        void setDisableErrorFCM(boolean en);

        boolean isDisabledErrorFCM();

        boolean isSettings_no_push();

        boolean isCommentsDesc();

        boolean toggleCommentsDirection();

        boolean isInfo_reading();

        boolean isAuto_read();

        boolean isNot_update_dialogs();

        boolean isBe_online();

        int getDonate_anim_set();

        int getColorChat();

        int getSecondColorChat();

        boolean isCustom_chat_color();

        int getColorMyMessage();

        int getSecondColorMyMessage();

        boolean isCustom_MyMessage();

        boolean isUse_stop_audio();

        boolean isPlayer_Has_Background();

        boolean isShow_mini_player();

        boolean isEnable_show_recent_dialogs();

        boolean is_side_navigation();

        boolean is_side_no_stroke();

        boolean is_side_transition();

        boolean is_notification_force_link();

        boolean isEnable_show_audio_top();

        boolean isUse_internal_downloader();

        boolean isEnable_last_read();

        boolean isNot_read_show();

        boolean isHeaders_in_dialog();

        String getMusicDir();

        String getPhotoDir();

        String getVideoDir();

        String getDocDir();

        String getStickerDir();

        boolean isPhoto_to_user_dir();

        boolean isDownload_voice_ogg();

        boolean isDelete_cache_images();

        boolean isCompress_traffic();

        boolean isDo_not_clear_back_stack();

        boolean isMention_fave();

        boolean isDisabled_encryption();

        boolean isDownload_photo_tap();

        boolean isDisable_sensored_voice();

        boolean isInvertPhotoRev();

        void setInvertPhotoRev(boolean rev);

        boolean isAudio_save_mode_button();

        boolean isShow_mutual_count();

        boolean isNot_friend_show();

        boolean isDo_zoom_photo();

        boolean isChange_upload_size();

        boolean isShow_photos_line();

        boolean isDisable_likes();

        void setDisable_likes(boolean disabled);

        boolean isDisable_notifications();

        void setDisable_notifications(boolean disabled);

        boolean isDo_auto_play_video();

        boolean isVideo_controller_to_decor();

        boolean isVideo_swipes();

        boolean isHint_stickers();

        boolean isEnable_native();

        boolean isEnable_cache_ui_anim();

        boolean isRecording_to_opus();

        int getPaganSymbol();

        boolean isRunes_show();

        int getMusicLifecycle();

        int getFFmpegPlugin();

        @Lang
        int getLanguage();

        int getRendering_mode();

        int getEndListAnimation();

        boolean appStoredVersionEqual();

        @NonNull
        LocalServerSettings getLocalServer();

        void setLocalServer(@NonNull LocalServerSettings settings);

        @NonNull
        PlayerCoverBackgroundSettings getPlayerCoverBackgroundSettings();

        void setPlayerCoverBackgroundSettings(@NonNull PlayerCoverBackgroundSettings settings);

        @NonNull
        SlidrSettings getSlidrSettings();

        void setSlidrSettings(@NonNull SlidrSettings settings);

        @Nullable
        String getUserNameChanges(int userId);

        void setUserNameChanges(int userId, @Nullable String name);

        void reloadUserNameChangesSettings(boolean onlyRoot);

        @NonNull
        Map<String, String> getUserNameChanges();

        @NonNull
        Set<String> getUserNameChangesKeys();

        int getCustomChannelNotif();

        void nextCustomChannelNotif();
    }

    interface IAccountsSettings {
        int INVALID_ID = -1;

        Flowable<Integer> observeChanges();

        Flowable<IAccountsSettings> observeRegistered();

        List<Integer> getRegistered();

        int getCurrent();

        void setCurrent(int accountId);

        void remove(int accountId);

        void registerAccountId(int accountId, boolean setCurrent);

        void storeAccessToken(int accountId, String accessToken);

        void storeLogin(int accountId, String loginCombo);

        void removeDevice(int accountId);

        void storeDevice(int accountId, String deviceName);

        @Nullable
        String getDevice(int accountId);

        String getLogin(int accountId);

        void storeTokenType(int accountId, @AccountType int type);

        String getAccessToken(int accountId);

        @AccountType
        int getType(int accountId);

        void removeAccessToken(int accountId);

        void removeType(int accountId);

        void removeLogin(int accountId);
    }

    interface IMainSettings {

        boolean isSendByEnter();

        boolean isMy_message_no_color();

        boolean isNotification_bubbles_enabled();

        boolean isMessages_menu_down();

        boolean isExpand_voice_transcript();

        @ThemeOverlay
        int getThemeOverlay();

        boolean isAudio_round_icon();

        boolean isUse_long_click_download();

        boolean isRevert_play_audio();

        boolean isShow_bot_keyboard();

        boolean isPlayer_support_volume();

        int isOpenUrlInternal();

        @Nullable
        Integer getUploadImageSize();

        void setUploadImageSize(Integer size);

        int getUploadImageSizePref();

        @PhotoSize
        int getPrefPreviewImageSize();

        void notifyPrefPreviewSizeChanged();

        @PhotoSize
        int getPrefDisplayImageSize(@PhotoSize int byDefault);

        @Transformers_Types
        int getViewpager_page_transform();

        @Transformers_Types
        int getPlayer_cover_transform();

        int getStart_newsMode();

        void setPrefDisplayImageSize(@PhotoSize int size);

        boolean isWebview_night_mode();

        boolean isSnow_mode();

        int getPhotoRoundMode();

        int getFontSize();

        boolean isLoad_history_notif();

        boolean isDont_write();

        boolean isOver_ten_attach();

        int cryptVersion();
    }

    interface INotificationSettings {
        int FLAG_SOUND = 1;
        int FLAG_VIBRO = 2;
        int FLAG_LED = 4;
        int FLAG_SHOW_NOTIF = 8;
        int FLAG_HIGH_PRIORITY = 16;

        int getNotifPref(int aid, int peerid);

        void setDefault(int aid, int peerId);

        void resetAccount(int aid);

        void forceDisable(int aid, int peerId);

        void setNotifPref(int aid, int peerid, int flag);

        int getOtherNotificationMask();

        boolean isCommentsNotificationsEnabled();

        boolean isFriendRequestAcceptationNotifEnabled();

        boolean isNewFollowerNotifEnabled();

        boolean isWallPublishNotifEnabled();

        boolean isGroupInvitedNotifEnabled();

        boolean isReplyNotifEnabled();

        boolean isNewPostOnOwnWallNotifEnabled();

        boolean isNewPostsNotificationEnabled();

        boolean isLikeNotificationEnable();

        Uri getFeedbackRingtoneUri();

        Uri getNewPostRingtoneUri();

        String getDefNotificationRingtone();

        String getNotificationRingtone();

        void setNotificationRingtoneUri(String path);

        long[] getVibrationLength();

        boolean isQuickReplyImmediately();

        boolean isBirthdayNotifyEnabled();

        boolean isMentionNotifyEnabled();

        boolean isSilentChat(int aid, int peerId);

        @NonNull
        Map<String, Integer> getChatsNotif();

        @NonNull
        Set<String> getChatsNotifKeys();

        void reloadNotifSettings(boolean onlyRoot);
    }

    interface IRecentChats {
        List<RecentChat> get(int acountid);

        void store(int accountid, List<RecentChat> chats);
    }

    interface IDrawerSettings {
        boolean isCategoryEnabled(@SwitchableCategory int category);

        void setCategoriesOrder(@SwitchableCategory int[] order, boolean[] active);

        int[] getCategoriesOrder();

        Observable<Object> observeChanges();
    }

    interface ISideDrawerSettings {
        boolean isCategoryEnabled(@SideSwitchableCategory int category);

        void setCategoriesOrder(@SideSwitchableCategory int[] order, boolean[] active);

        int[] getCategoriesOrder();

        Observable<Object> observeChanges();
    }

    interface IPushSettings {
        void savePushRegistations(Collection<VkPushRegistration> data);

        List<VkPushRegistration> getRegistrations();
    }

    interface ISecuritySettings {
        boolean isKeyEncryptionPolicyAccepted();

        void setKeyEncryptionPolicyAccepted(boolean accepted);

        boolean isPinValid(@NonNull int[] values);

        void setPin(@Nullable int[] pin);

        boolean isUsePinForEntrance();

        boolean isUsePinForSecurity();

        boolean isEntranceByFingerprintAllowed();

        @KeyLocationPolicy
        int getEncryptionLocationPolicy(int accountId, int peerId);

        void disableMessageEncryption(int accountId, int peerId);

        boolean isMessageEncryptionEnabled(int accountId, int peerId);

        void enableMessageEncryption(int accountId, int peerId, @KeyLocationPolicy int policy);

        void firePinAttemptNow();

        void clearPinHistory();

        List<Long> getPinEnterHistory();

        boolean hasPinHash();

        int getPinHistoryDepth();

        boolean needHideMessagesBodyForNotif();

        void addHiddenDialog(int peerId);

        void removeHiddenDialog(int peerId);

        boolean hasHiddenDialogs();

        boolean isHiddenDialog(int peerId);

        boolean getShowHiddenDialogs();

        void setShowHiddenDialogs(boolean showHiddenDialogs);

        void reloadHiddenDialogSettings();

        boolean isDelayedAllow();

        void updateLastPinTime();

        boolean IsShow_hidden_accounts();
    }

    interface IUISettings {
        void setMainTheme(String key);

        void switchNightMode(@NightMode int key);

        String getMainThemeKey();

        @AvatarStyle
        int getAvatarStyle();

        void storeAvatarStyle(@AvatarStyle int style);

        boolean isDarkModeEnabled(Context context);

        int getNightMode();

        Place getDefaultPage(int accountId);

        void notifyPlaceResumed(int type);

        boolean isSystemEmoji();

        boolean isEmojis_full_screen();

        boolean isStickers_by_theme();

        boolean isStickers_by_new();

        boolean isShow_profile_in_additional_page();

        @SwipesChatMode
        int getSwipes_chat_mode();

        boolean isDisplay_writing();
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/settings/ISettings.kt
package dev.ragnarok.fenrir.settings

import android.content.Context
import androidx.appcompat.app.AppCompatDelegate
import dev.ragnarok.fenrir.AccountType
import dev.ragnarok.fenrir.api.model.LocalServerSettings
import dev.ragnarok.fenrir.api.model.PlayerCoverBackgroundSettings
import dev.ragnarok.fenrir.api.model.SlidrSettings
import dev.ragnarok.fenrir.crypt.KeyLocationPolicy
import dev.ragnarok.fenrir.model.DrawerCategory
import dev.ragnarok.fenrir.model.Lang
import dev.ragnarok.fenrir.model.PhotoSize
import dev.ragnarok.fenrir.model.drawer.RecentChat
import dev.ragnarok.fenrir.place.Place
import dev.ragnarok.fenrir.settings.theme.ThemeOverlay
import dev.ragnarok.fenrir.view.pager.Transformers_Types
import kotlinx.coroutines.flow.SharedFlow

interface ISettings {
    fun recentChats(): IRecentChats
    fun drawerSettings(): IDrawerSettings
    fun sideDrawerSettings(): ISideDrawerSettings
    fun pushSettings(): IPushSettings
    fun security(): ISecuritySettings
    fun ui(): IUISettings
    fun notifications(): INotificationSettings
    fun main(): IMainSettings
    fun accounts(): IAccountsSettings

    interface IAccountsSettings {
        val observeChanges: SharedFlow<Long>
        val observeRegistered: SharedFlow<IAccountsSettings>
        val registered: List<Long>
        var current: Long
        val currentAccessToken: String?

        @AccountType
        val currentType: Int
        val currentHidden: Boolean
        fun remove(accountId: Long): Long?
        fun registerAccountId(accountId: Long, setCurrent: Boolean)
        fun storeAccessToken(accountId: Long, accessToken: String?)
        fun storeLogin(accountId: Long, loginCombo: String?)
        fun removeDevice(accountId: Long)
        fun storeDevice(accountId: Long, deviceName: String?)
        fun getDevice(accountId: Long): String?
        fun getLogin(accountId: Long): String?
        fun storeTokenType(accountId: Long, @AccountType type: Int)
        fun getAccessToken(accountId: Long): String?

        @AccountType
        fun getType(accountId: Long): Int
        fun removeAccessToken(accountId: Long)
        fun removeType(accountId: Long)
        fun removeLogin(accountId: Long)

        fun loadAccounts(refresh: Boolean)

        var anonymToken: AnonymToken

        companion object {
            const val INVALID_ID = -1L
        }
    }

    interface IMainSettings {
        val isSendByEnter: Boolean
        val isNotification_bubbles_enabled: Boolean
        val isMessages_menu_down: Boolean
        val isExpand_voice_transcript: Boolean
        val isChat_popup_menu: Boolean

        @get:ThemeOverlay
        val themeOverlay: Int
        val isAudio_round_icon: Boolean
        val isUse_long_click_download: Boolean
        val isRevert_play_audio: Boolean
        val isShow_bot_keyboard: Boolean
        val isPlayer_support_volume: Boolean
        val isOpenUrlInternal: Int
        var uploadImageSize: Int?
        val uploadImageSizePref: Int

        @get:PhotoSize
        val prefPreviewImageSize: Int
        fun notifyPrefPreviewSizeChanged()

        @PhotoSize
        fun getPrefDisplayImageSize(@PhotoSize byDefault: Int): Int

        @get:Transformers_Types
        val viewpager_page_transform: Int

        @get:Transformers_Types
        val player_cover_transform: Int
        val start_newsMode: Int
        fun setPrefDisplayImageSize(@PhotoSize size: Int)
        val isWebview_night_mode: Boolean
        val single_line_videos: Boolean
        val single_line_photos: Boolean
        val isSnow_mode: Boolean
        val photoRoundMode: Int
        val fontSize: Int
        val fontSizeOnlyForChatsAndMessages: Boolean
        val isLoad_history_notif: Boolean
        val isDont_write: Boolean
        val isOver_ten_attach: Boolean
        val cryptVersion: Int

        fun getFeedSourceIds(accountId: Long): String?
        fun setFeedSourceIds(accountId: Long, sourceIds: String?)
        fun storeFeedScrollState(accountId: Long, state: String?)
        fun restoreFeedScrollState(accountId: Long): String?
        fun restoreFeedNextFrom(accountId: Long): String?
        fun storeFeedNextFrom(accountId: Long, nextFrom: String?)
        val isAudioBroadcastActive: Boolean
        val maxBitmapResolution: Int
        val servicePlaylist: List<Int>
        val isValidate_tls: Boolean
        val isNative_parcel_photo: Boolean
        val isNative_parcel_story: Boolean
        val isNative_parcel_docs: Boolean
        val isDoLogs: Boolean
        val isDump_fcm: Boolean
        val isAutoplay_video_on_posts: Int
        val isStrip_news_repost: Boolean
        val isCommunities_in_page_search: Boolean
        val isAd_block_story_news: Boolean
        val isBlock_news_by_words: Set<String>?
        val isNew_loading_dialog: Boolean
        val apiDomain: String
        val authDomain: String
        val isUse_api_5_90_for_audio: Boolean
        val isSaving_network_traffic: Boolean
        val isDisable_history: Boolean
        val isShow_wall_cover: Boolean
        val isDeveloper_mode: Boolean
        val isForce_cache: Boolean
        val isKeepLongpoll: Boolean
        fun setDisableErrorFCM(en: Boolean)
        val isDisabledErrorFCM: Boolean
        val isSettings_no_push: Boolean
        val isCommentsDesc: Boolean
        val toggleCommentsDirection: Boolean
        val isInfo_reading: Boolean
        val isAuto_read: Boolean
        val isMarkListenedVoice: Boolean
        val isNot_update_dialogs: Boolean
        val isBe_online: Boolean
        val donate_anim_set: Int
        val colorChat: Int
        val secondColorChat: Int
        val isCustom_chat_color: Boolean

        val isMessageOutNoColor: Boolean
        val isCustomMessageOutColor: Boolean
        val customColorMessageOutPrimary: Int
        val customColorMessageOutSecondary: Int

        val isCustomMessageInColor: Boolean
        val customColorMessageInPrimary: Int
        val customColorMessageInSecondary: Int

        val isUse_stop_audio: Boolean
        val isPlayer_Has_Background: Boolean
        val isShow_mini_player: Boolean
        val isEnable_show_recent_dialogs: Boolean
        val is_side_navigation: Boolean
        val is_side_no_stroke: Boolean
        val is_side_transition: Boolean
        val is_notification_force_link: Boolean
        val isEnable_show_audio_top: Boolean
        val isUse_internal_downloader: Boolean
        val isEnable_last_read: Boolean
        val isNot_read_show: Boolean
        val isHeaders_in_dialog: Boolean
        val musicDir: String
        val photoDir: String
        val videoDir: String
        val docDir: String
        val stickerDir: String
        val isPhoto_to_user_dir: Boolean
        val isDownload_voice_ogg: Boolean
        val isDelete_cache_images: Boolean
        val isCompress_incoming_traffic: Boolean
        val isCompress_outgoing_traffic: Boolean
        val isLimitImage_cache: Int
        val isDo_not_clear_back_stack: Boolean
        val isMention_fave: Boolean
        val isDisabled_encryption: Boolean
        val isDisable_sensored_voice: Boolean
        var isInvertPhotoRev: Boolean
        val isAudio_save_mode_button: Boolean
        val isShow_mutual_count: Boolean
        val isDo_zoom_photo: Boolean
        val isChange_upload_size: Boolean
        val isInstant_photo_display: Boolean
        val isShow_photos_line: Boolean
        val isShow_photos_date: Boolean
        var isDisable_likes: Boolean
        var isDisable_notifications: Boolean
        val isDo_auto_play_video: Boolean
        val isVideo_controller_to_decor: Boolean
        val isVideo_swipes: Boolean
        val isHint_stickers: Boolean
        val isEnable_native: Boolean
        val isRecording_to_opus: Boolean
        val paganSymbol: Int
        val isRunes_show: Boolean
        val musicLifecycle: Int
        val fFmpegPlugin: Int
        var isRememberLocalAudioAlbum: Boolean
        var currentLocalAudioAlbum: Int
        val videoExt: Set<String>
        val photoExt: Set<String>
        val audioExt: Set<String>
        val maxThumbResolution: Int
        val isPhoto_zoom_enable_list: Boolean
        val longClickPhoto: Int
        val isEnable_dirs_files_count: Boolean
        val last_audio_sync: Long
        fun set_last_audio_sync(time: Long)
        fun get_last_sticker_sets_sync(accountId: Long): Long
        fun set_last_sticker_sets_sync(accountId: Long, time: Long)
        fun get_last_sticker_sets_custom_sync(accountId: Long): Long
        fun set_last_sticker_sets_custom_sync(accountId: Long, time: Long)
        fun get_stickers_version_hash(accountId: Long): String?
        fun set_stickers_version_hash(accountId: Long, hash: String?)
        fun del_last_sticker_sets_sync(accountId: Long)
        fun del_last_sticker_sets_custom_sync(accountId: Long)
        fun del_stickers_version_hash(accountId: Long)
        fun del_last_reaction_assets_sync(accountId: Long)
        fun set_last_reaction_assets_sync(accountId: Long, time: Long)
        fun get_last_reaction_assets_sync(accountId: Long): Long
        fun reloadOwnerChangesMonitor()
        fun isOwnerInChangesMonitor(ownerId: Long): Boolean
        fun putOwnerInChangesMonitor(ownerId: Long)
        fun removeOwnerInChangesMonitor(ownerId: Long)
        fun resetAllChangesMonitor()
        val isAudio_catalog_v2: Boolean

        val picassoDispatcher: Int

        @get:Lang
        val language: Int
        val rendering_mode: Int
        val endListAnimation: Int
        val appStoredVersionEqual: Boolean
        var localServer: LocalServerSettings
        var playerCoverBackgroundSettings: PlayerCoverBackgroundSettings
        var slidrSettings: SlidrSettings
        fun getUserNameChanges(userId: Long): String?
        fun setUserNameChanges(userId: Long, name: String?)
        fun reloadUserNameChangesSettings(onlyRoot: Boolean)
        val userNameChangesMap: Map<String, String>
        fun resetAllUserNameChanges()
        val customChannelNotif: Int
        fun nextCustomChannelNotif()
        val currentParser: Int
        var catalogV2ListSort: List<Int>
        val isOnlyNotViewedFollowers: Boolean
    }

    interface INotificationSettings {
        fun resetAccount(aid: Long)
        fun resetAll()

        fun setSilentPeer(aid: Long, peerId: Long, silent: Boolean)
        fun isSilentPeer(aid: Long, peerId: Long): Boolean

        val silentPeersMap: Map<String, Boolean>

        fun reloadSilentSettings(onlyRoot: Boolean)
    }

    interface IRecentChats {
        operator fun get(accountId: Long): MutableList<RecentChat>
        fun store(accountId: Long, chats: List<RecentChat>)
    }

    interface IDrawerSettings {
        var categoriesOrder: List<DrawerCategory>
        val observeChanges: SharedFlow<List<DrawerCategory>>
        fun reset()
    }

    interface ISideDrawerSettings {
        var categoriesOrder: List<DrawerCategory>
        val observeChanges: SharedFlow<List<DrawerCategory>>
        fun reset()
    }

    interface IPushSettings {
        var registered: VKPushRegistration?
    }

    interface ISecuritySettings {
        var isKeyEncryptionPolicyAccepted: Boolean
        fun isPinValid(values: IntArray): Boolean
        fun setPin(pin: IntArray?)
        var isUsePinForSecurity: Boolean
        var isUsePinForEntrance: Boolean
        var isDelayedAllow: Boolean
        var isEntranceByFingerprintAllowed: Boolean

        @KeyLocationPolicy
        fun getEncryptionLocationPolicy(accountId: Long, peerId: Long): Int
        fun disableMessageEncryption(accountId: Long, peerId: Long)
        fun isMessageEncryptionEnabled(accountId: Long, peerId: Long): Boolean
        fun enableMessageEncryption(accountId: Long, peerId: Long, @KeyLocationPolicy policy: Int)
        fun firePinAttemptNow()
        fun clearPinHistory()
        val pinEnterHistory: List<Long>
        val hasPinHash: Boolean
        val pinHistoryDepthValue: Int
        val needHideMessagesBodyForNotif: Boolean
        fun addHiddenDialog(peerId: Long)
        fun removeHiddenDialog(peerId: Long)
        val hasHiddenDialogs: Boolean
        fun isHiddenDialog(peerId: Long): Boolean
        var showHiddenDialogs: Boolean
        fun reloadHiddenDialogSettings()
        fun updateLastPinTime()
        val IsShow_hidden_accounts: Boolean
    }

    interface IUISettings {
        fun setMainTheme(key: String)
        fun switchNightMode(@AppCompatDelegate.NightMode key: Int)
        val mainThemeKey: String

        @get:AvatarStyle
        val avatarStyle: Int
        fun storeAvatarStyle(@AvatarStyle style: Int)
        fun isDarkModeEnabled(context: Context): Boolean
        val nightMode: Int
        fun getDefaultPage(accountId: Long): Place
        fun notifyPlaceResumed(type: Int)
        val isEmojis_full_screen: Boolean
        val isStickers_by_theme: Boolean
        val isStickers_by_new: Boolean
        val isShow_profile_in_additional_page: Boolean

        @get:SwipesChatMode
        val swipes_chat_mode: Int
        val isDisplay_writing: Boolean
    }
}